File: code/boj_25200.py
'''
    곰곰이와 자판기
    https://www.acmicpc.net/problem/25200
'''
N,M=map(int,input().split())
values = []
for i in range(N):
  node = [i, None]
  values.append([node,node])

for _ in range(M):
  u, v = map(int,input().split())
  u -= 1
  v -= 1

  if values[u][0] == None: continue
  if values[v][0] != None:
    values[u][1][1] = values[v][0]
    values[u] = [values[u][0], values[v][1]]

  values[v] = values[u]
  values[u] = [None, None]

result = [0]*N
for i in range(N):
  it = values[i][0]
  while it != None:
    result[it[0]] = i+1
    it = it[1]

print(*result)

File: code/boj_25200.cpp
/*
    곰곰이와 자판기
    https://www.acmicpc.net/problem/25200
*/
#include <bits/stdc++.h>

using namespace std;

int main()
{
	cin.tie(nullptr)->sync_with_stdio(false);
	size_t N;
	cin >> N;

	vector<size_t> value(N);
	vector<size_t> parent(N);
	vector<size_t> index(N);
	for(size_t i=0; i!=N; ++i) index[i] = value[i] = parent[i] = i;
	function<size_t(size_t)> root = [&](size_t n){
		if(parent[n] == n) return n;
		return parent[n] = root(parent[n]);
	};

	auto merge = [&](size_t u, size_t v){
		if(index[u] == -1) return;

		size_t before = u;
		u = root(index[u]);
		index[before] = -1;
		value[u] = v;
		if(index[v] == -1)
			index[v] = u;
		else
			parent[u] = root(index[v]);
	};

	size_t M;
	cin >> M;
	while(M--){
		size_t u, v;
		cin >> u >> v;
		--u; --v;

		merge(u, v);
	}

	for(size_t i=0; i!=N; ++i)
		cout << value[root(i)]+1 << " ";
}